### DISCUSSION TITLE
expanding environment variables on Vercel

### DISCUSSION THREAD

**Discussion by joggienl** [vercel/vercel#8579] [Help] [answered]
Does anybody know if you can setup vervel environment variables to use other variables?

As an example of what I try to achieve, could I use the `VERCEL_URL` environment variable in a custom environment variable:

<img width="988" alt="image" src="https://user-images.githubusercontent.com/1089150/191020168-325ce06a-a02a-44af-a470-1bfddc7ba6ea.png">

This does work locally in your .env.local file, but it does not seem to work on Vercel.

**Comment by joggienl** [comment]
The nextjs documentation has what I need but apparently this does not work on Vercel. Ideas are welcome 😅 

> Note: Next.js will automatically expand variables ($VAR) inside of your .env* files. This allows you to reference other secrets, like so:
> 
```
# .env
HOSTNAME=localhost
PORT=8080
HOST=http://$HOSTNAME:$PORT
```

If you are trying to use a variable with a `$` in the actual value, it needs to be escaped like so: `\$`.

>
>For example:
```
# .env
A=abc

# becomes "preabc"
WRONG=pre$A

# becomes "pre$A"
CORRECT=pre\$A
```

Source: [environment variable expansion ](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables)

**Reply by amyegan** [reply]
Thanks for sharing this @joggienl! I'm asking the team about this. Someone with more background on this project will be able to give us more info.

**Reply by joggienl** [reply]
Thanks @amyegan 

I also came across this issue: https://github.com/vercel/next.js/discussions/16429

I tried to create a `.env.production` file. In that file expanding variables works as expected.

```
# .env.production

NEXT_PUBLIC_SOME_VAR=https://$VERCEL_URL/some/fake/path
```

When deployed, `process.env.NEXT_PUBLIC_SOME_VAR` has the value I would suspect. So it actually just does not work from the Vercel dashboard.

**Reply by amyegan** [reply]
I think you figured it out @joggienl. Confirmed with the engineers that Next.js can expand variables from `.env` in source code, but it cannot with variables in the dashboard.

**Reply by joggienl** [reply]
Cool, good to know. Perhaps it is good to also update the NextJS and Vercel documentation that it only works in your .env.* files? That is not really obvious right know.

I am not sure if it is worth to be able to do this from the dashboard. Only reasons I can come up with is that you might not want production environment information in your codebase and it might be confusing if you have both an `.env.production` _and_ setup via the dashboard. Which is inevitable imo.

Any thoughts?

**Reply by amyegan** [reply]
I agree. I'm adding a task to update the documentation and asking the team take a closer look at this feature. Thanks for bringing this up!

**Reply by joggienl** [reply]
Thanks @amyegan !

**Reply by T99** [reply]
@amyegan Any updates/timeline on when this might be fixed?

**Reply by iamarpitpatidar** [reply]
@amyegan any updates on the feature yet ?

**Reply by julietgar** [reply]
@amyegan any update with this? I'm interested on this feature. Thanks!

**Reply by Wazbat** [reply]
Any update on this?

**Comment by amyegan** [comment, marked as answer]
Hey 👋 The [original solution](https://github.com/vercel/vercel/discussions/8579#discussioncomment-3697826) is still valid, but I want to share an update with links to some related docs. Next.js, and some other frameworks, will automatically expand `$` variables in your `.env*` files.

- [Referencing Other Variables](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#referencing-other-variables)
- [Environment Variable Load Order](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#environment-variable-load-order)

There are likely many possible ways to reach the same goal. If the [recommended strategy](https://github.com/vercel/next.js/discussions/16429#discussioncomment-1302156) doesn't work for your particular app or workflow, you can try some of the other ideas shared in https://github.com/vercel/next.js/discussions/16429. If you find a different solution, let us know what worked best for you!
